### PR TITLE
Fix UnaliasSymbolReferences

### PR DESCRIPTION
**Pull request by aaneja**
For arguments of non-deterministic assignments. These were incorrectly
being canonicalized causing a subtle bug in how non-determinstic
functions are computed

Fixes https://github.com/prestodb/presto/issues/23318

## Test Plan
New test added

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by elharo — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 1]
```diff
@@ -131,33 +134,76 @@ public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider
...
     }
 
+    private static class NonDeterministicSymbolsExtractor
+            extends SimplePlanVisitor<ImmutableSet.Builder<VariableReferenceExpression>>
+    {
+        private final RowExpressionDeterminismEvaluator determinismEvaluator;
+
+        public NonDeterministicSymbolsExtractor(RowExpressionDeterminismEvaluator determinismEvaluator)
+        {
+            this.determinismEvaluator = determinismEvaluator;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:154]:
```suggestion
            this.determinismEvaluator = requireNonNull(determinismEvaluator, "determinismEvaluator is null");
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 2]
```diff
@@ -131,33 +134,76 @@ public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider
         requireNonNull(types, "types is null");
         requireNonNull(variableAllocator, "variableAllocator is null");
         requireNonNull(idAllocator, "idAllocator is null");
-        PlanNode rewrittenPlan = SimplePlanRewriter.rewriteWith(new Rewriter(types, functionAndTypeManager, warningCollector), plan);
+
+        NonDeterministicSymbolsExtractor nonDeterministicSymbolsExtractor = new NonDeterministicSymbolsExtractor(new RowExpressionDeterminismEvaluator(functionAndTypeManager));
+        ImmutableSet.Builder<VariableReferenceExpression> builder = ImmutableSet.builder();
+        plan.accept(nonDeterministicSymbolsExtractor, builder);
+        ImmutableSet<VariableReferenceExpression> doNotCanonicalize = builder.build();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:141]: Perhaps extract this into a method, `getNonDeterministicSymbols`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 3]
```diff
@@ -694,6 +740,10 @@ public PlanNode visitPlan(PlanNode node, RewriteContext<Void> context)
         private void map(VariableReferenceExpression variable, VariableReferenceExpression canonical)
         {
             Preconditions.checkArgument(!variable.equals(canonical), "Can't map variable to itself: %s", variable);
+            if (doNotCanonicalize.contains(variable)) {
+                return;
+            }
+
             mapping.put(variable.getName(), canonical.getName());
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:747]:
```suggestion
            if (!doNotCanonicalize.contains(variable)) {
                mapping.put(variable.getName(), canonical.getName());
            }
```

**Review by tdcmeehan — COMMENTED** [review]
Just some nits

**Comment by aaneja** [comment]
Thanks for linking this to the legacy discussion on this (long standing) issue. From what I could gather, we don't have consensus on how we want the behavior of 'merging non-deterministic functions' addressed
(Should we be doing *any* merging ? What cases does it make sense ?)

I will close this PR until we can drive consensus on this

**Comment by aaneja** [comment]
Just to clarify -

The change I was proposing here is that we do not 'un-alias' any symbols in any assignments if they are used as arguments to any non-deterministic expressions.
Right now, the change is very broad - I am not performing *any* unaliasing for these symbols if they are used in any non-deterministic context.  
I think scoping this down to just the non-deterministic functions would be possible

This does stop the 'merge' of the `random()` calls, but does so by avoiding building duplicate assigments in the first place

**Pull request closed without merging**